contracts/p1/Main.sol

```diff
@@ -58,14 +92,74 @@ contract MainP1 is Versioned, Initializable, Auth, ComponentRegistry, UUPSUpgrad
         return super.hasRole(role, account);
     }
 
+    /**
+     * @dev When upgrading from a prior version to 4.0.0,
+     *      this must happen in the Governance proposal.
+     */
+    function upgradeMainTo(bytes32 versionHash) external onlyRole(OWNER) {
+        require(address(versionRegistry) != address(0), "no registry");
+        require(!versionRegistry.isDeprecated(versionHash), "version deprecated");
+
+        Implementations memory implementation = versionRegistry.getImplementationForVersion(
+            versionHash
+        );
+
+        this.upgradeTo(address(implementation.main));
+    }
+
+    function upgradeRTokenTo(bytes32 versionHash, bool validateProposal) external onlyRole(OWNER) {
+        require(address(versionRegistry) != address(0), "no registry");
+        require(keccak256(abi.encodePacked(this.version())) == versionHash, "upgrade main first");
+
+        Implementations memory implementation = versionRegistry.getImplementationForVersion(
+            versionHash
+        );
+
+        if (validateProposal) {
```

Comment:
can't we assume that the registry is valid before the upgrade? or is this just to cover the case where the RToken is initially upgrading to 4.0.0?

Reply:
It's for _any_ upgrade including from/to 4.0.0, hence it's optional in both directions.

Reply:
It does seem like we need to pull apart before and after the upgrade

Reply:
imo:
- required after the proposal
- optional before